src/webgpu/api/validation/queue/copyImageBitmapToTexture.spec.ts

```diff
@@ -0,0 +1,409 @@
+  switch (dimension) {
+    case '1d':
+      return poptions('dstTextureSize', [
+        { width: DEFAULT_WIDTH, height: 1, depth: 1 },
+        { width: DEFAULT_WIDTH + 1, height: 1, depth: 1 },
+      ]);
+    case '2d':
+      return poptions('dstTextureSize', [
+        { width: DEFAULT_WIDTH, height: DEFAULT_HEIGHT, depth: 1 },
+        { width: DEFAULT_WIDTH + 1, height: DEFAULT_HEIGHT + 1, depth: 1 },
+      ]);
+    case '3d':
+      return poptions('dstTextureSize', [
+        { width: DEFAULT_WIDTH, height: DEFAULT_HEIGHT, depth: DEFAULT_DEPTH },
+        { width: DEFAULT_WIDTH + 1, height: DEFAULT_HEIGHT + 1, depth: DEFAULT_DEPTH + 1 },
+      ]);
+  }
+}
+
+// Generate source origin value[valid, invalid] which are GPUOrigin2D type.
+function generateSrcOriginValue() {
+  const value = [];
+  // Values for each dimensions are valid or too large.
+  for (let i = 0; i <= DEFAULT_WIDTH; i += DEFAULT_WIDTH) {
+    for (let j = 0; j <= DEFAULT_HEIGHT; j += DEFAULT_HEIGHT) {
+      value.push({ x: i, y: j });
+    }
+  }
+
+  return value;
+}
+
+// Generate dst origin value[valid, invalid] which are GPUOrigin3D type
+function generateDstOriginValue() {
+  const value = [];
+  // Values for each dimensions are valid or too large.
+  for (let i = 0; i <= DEFAULT_WIDTH; i += DEFAULT_WIDTH) {
+    for (let j = 0; j <= DEFAULT_HEIGHT; j += DEFAULT_HEIGHT) {
+      for (let k = 0; k <= DEFAULT_DEPTH; k += DEFAULT_DEPTH) {
+        value.push({ x: i, y: j, z: k });
+      }
+    }
+  }
+
+  return value;
+}
+
+// Generate copySizes[valid, invalid], which are GPUExtent3D type
+function generateCopySizeValue() {
+  const value = [];
+  for (let i = 0; i <= DEFAULT_WIDTH; i += DEFAULT_WIDTH) {
+    for (let j = 0; j <= DEFAULT_HEIGHT; j += DEFAULT_HEIGHT) {
+      for (let k = 0; k <= DEFAULT_DEPTH; k += DEFAULT_DEPTH) {
+        value.push({ width: i, height: j, depth: k });
+      }
+    }
+  }
+  return value;
+}
+
+class CopyImageBitmapToTextureTest extends ValidationTest {
+  private getImageData(width: number, height: number): ImageData {
+    const pixelSize = DEFAULT_BYTES_PER_PIXEL * width * height;
+    const imagePixels = new Uint8ClampedArray(pixelSize);
+    for (let i = 0; i < pixelSize; ++i) {
+      imagePixels[i] = i;
+    }
+    return new ImageData(imagePixels, width, height);
+  }
+
+  async getImageBitmapCopyView(
+    width: number,
+    height: number,
+    origin: GPUOrigin2D,
+    closed: boolean
+  ): Promise<GPUImageBitmapCopyView> {
+    const imageBitmap = await createImageBitmap(this.getImageData(width, height));
+    if (closed) {
+      imageBitmap.close();
+    }
+    return { imageBitmap, origin };
+  }
+
+  // Get default source copy view with Default width, height and correct origin.
+  async getDefaultImageBitmapCopyView(): Promise<GPUImageBitmapCopyView> {
+    return this.getImageBitmapCopyView(DEFAULT_WIDTH, DEFAULT_HEIGHT, { x: 0, y: 0 }, false);
+  }
+
+  getTextureCopyView(
```

Comment:
I'm not sure about all these chained helpers, they don't seem to save so much typing in the tests (especially if taking advantage of the default values in descriptors) and make it harder to know in the test what values are used.

Reply:
You're right, I'll remove the chain helpers.